Add ClickHouse setting to disable tunneling for HTTPS requests over HTTP proxy

arthurpassos opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->

**Motivation**: It is very common to use a HTTP proxy to route HTTPS requests. The default way of doing that is through HTTP CONNECT, which establishes a tunnel with the remote server and the communication remains secure and encrypted. The proxy server does not see the content, which comes with the downside of: no caching. Some users trust their network and are willing to take the risk of sending unencrypted data to the proxy server to get the benefit of caching.

**New feature**: Add a setting under the proxy tag which allows disabling tunneling for https requests over http proxy. Default behavior (when not specified) is to use CONNECT protocol when the request protocol is HTTPS and the proxy is HTTP.

**Example**:
```
<clickhouse>
    <proxy>
        <disable_tunneling_for_https_requests_over_http_proxy>1</disable_tunneling_for_https_requests_over_http_proxy>
    </proxy>
<clickhouse>
```

### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ClickHouse setting to disable tunneling for HTTPS requests over HTTP proxy.

### Documentation entry for user-facing changes

- [ x ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

arthurpassos committed:
initial commit. integ tests passing, need to re-run unit & my own personal tests

arthurpassos commented on src/Common/ProxyConfigurationResolver.h:
```diff
@@ -9,9 +9,41 @@ struct ProxyConfigurationResolver
     virtual void errorReport(const ProxyConfiguration & config) = 0;
+
+protected:
+    Protocol request_protocol;
```
I feel bad for introducing state in this class (which used to be a simple interface), but the request protocol as well as connect protocol policy is required to determine if connect should be on or off.

arthurpassos replied:
This could actually be avoided for both environment resolver and remote resolver because the proxy protocol is known beforehand. When `ProxyConfigurationResolverProvider` is picking the configuration, it knows the proxy protocol and the request protocol, so it could simply inject a boolean.

The problem lies on `ProxyListConfigurationResolver` as it contains a list of URIs and that list might contain mixed proxy protocols. One thing I could do is have something like `std::vector<std::pair<URI, bool>>` instead of `std::vector<URI>` only for the list. That would simplify the other cases, but make this one a bit more complex.

Not sure it's worth it.

I am writing this comment (and code) with the feeling that there is a simpler way to address this

arthurpassos commented on src/Common/ProxyConfigurationResolver.h:
```diff
@@ -9,9 +9,41 @@ struct ProxyConfigurationResolver
+protected:
+    Protocol request_protocol;
+
+    // TODO think about any case
```
`Protocol::ANY` is only used for old settings format. When parsing old settings format, `use_connect_protocol` is not read. Therefore, it is always `ConnectProtocolPolicy::DEFAULT`.

arthurpassos replied:
Hm.. that might be a problem?

arthurpassos committed:
partial refactoring to remove Protocol::ANY

arthurpassos commented on src/Common/ProxyConfigurationResolverProvider.cpp:
```diff
@@ -227,7 +234,7 @@ std::shared_ptr<ProxyConfigurationResolver> ProxyConfigurationResolverProvider::
      * Falls back to Environment resolver if no configuration is found.
      * */
-    return ProxyConfigurationResolverProvider::get(Protocol::ANY, configuration);
+    return ProxyConfigurationResolverProvider::get(request_protocol, configuration);
```
This fixes a bug. We were supposed to be flexible only on the old settings format, but when fallback to new settings, it shouldn't be flexible.

arthurpassos committed:
improve naming

arthurpassos committed:
remove all usages of ProxyConfiguration::Protocol::ANY

arthurpassos committed:
fix ut

arthurpassos commented:
Next steps are:

1. Improve current testing coverage & find out how and if it's necessary to add integration tests for new functionality;
2. Test in real life scenarios

arthurpassos commented:
@tavplubix can you add the "can be tested" label?

arthurpassos commented:
This PR also removes `Protocol::ANY`, that's why there are some "unrelated" changes I guess.

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #55033 from arthurpassos/ClickHouse:add_proxy_connect_protocol_setting -->
*This is an automated comment for commit de3c831b3331711ec2b5519fe4cc60baae92fa4c with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55033/de3c831b3331711ec2b5519fe4cc60baae92fa4c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

arthurpassos committed:
blabla

arthurpassos committed:
Merge branch 'master' into add_proxy_connect_protocol_setting

arthurpassos committed:
support url functions as well

arthurpassos commented:
It ain't that simple to implement this. I just found out `Poco::HTTPClientSession::ProxyConfig::tunnel` does not work properly for all cases. Not only that, but it does not even exist in the original `Poco` project. It was implemented in a ClickHouse fork, so it's probably only half-way implemented.

arthurpassos committed:
support for HTTPS requests over HTTP proxy with tunneling off

arthurpassos committed:
remove gtestabc

arthurpassos committed:
fix silly mistake

arthurpassos committed:
Merge branch 'master' into add_proxy_connect_protocol_setting

arthurpassos committed:
...

arthurpassos committed:
remove usages of httpclientsession::proxyconfig in src/

arthurpassos committed:
got you

arthurpassos committed:
remove stale comment

arthurpassos committed:
it seems like I need reasonable defaults

arthurpassos committed:
fix ut

arthurpassos committed:
add some comments

arthurpassos committed:
remove no longer needed header

arthurpassos committed:
Merge branch 'master' into add_proxy_connect_protocol_setting

arthurpassos committed:
matrix out

arthurpassos committed:
add https over http proxy with no tunneling

arthurpassos committed:
Merge branch 'master' into add_proxy_connect_protocol_setting

arthurpassos committed:
soem docs

arthurpassos committed:
partial refactoring

arthurpassos committed:
rename to use_tunneling_for_https_requests_over_http_proxy

arthurpassos committed:
improve docs

arthurpassos committed:
use shorter version

arthurpassos committed:
remove useless test

arthurpassos commented:
Tunneling / CONNECT is only used when a HTTPS requests needs to be routed over a HTTP proxy server. That is the default way of doing it and most HTTP clients do that.

There is no other default proxy connection that requires it, thus it makes sense to scope this setting to that case (i.e, `use_tunneling_for_https_requests_over_http_proxy` instead of `use_tunneling`).

In order to disable this, it was required to:
1. Add a ClickHouse setting that allows this customization;
2. Save original `request protocol`. This is required to build the proper `HTTP GET` message. The implementation prior to this PR would assume it's always HTTP (https://github.com/ClickHouse/ClickHouse/blob/c0482cb537581ccbec1e6992aa179504624df19d/base/poco/Net/src/HTTPClientSession.cpp#L421), which is not correct in this case.
3. Check if proxy server is HTTP, protocol request is HTTPS and setting is off (https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-c064ca4cd5e4eb5d1014cba51c9781b25457941c7aa0e19a84bdbf4505102823R165). In this case, a HTTP session has to be created even though the final destination is HTTPS. That is because ClickHouse should communicate with the proxy only, which is HTTP. **This was the tricky part to get right**

This PR assumes `Poco::HTTPClientSession::ProxyConfig::tunnel` being set to true or false will only impact HTTPS requests over HTTP proxy. According to my tests and debugging, this is true. For instance, proxy tunnel was being set to true or false based on that condition. Plus, it defaults to true. https://github.com/ClickHouse/ClickHouse/blob/e25ce7c5fbe73b0a9067805d19f43a27983e2973/src/IO/HTTPCommon.cpp#L199

arthurpassos commented:
@tavplubix hi again :)

arthurpassos commented:
Messed up, need to fix it, glad I had tests :)

arthurpassos committed:
rename the setting

arthurpassos committed:
update

arthurpassos committed:
fix typo

arthurpassos commented:
CI is as green as it can be, first time seeing this :eyes:

arthurpassos commented:
Hi @pufit, any news on this one?

pufit commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -2741,3 +2741,7 @@ Proxy settings are determined in the following order:
 it will check the next highest priority resolver type, until it reaches the environment resolver.
 This also allows a mix of resolver types can be used.
+
+### disable_tunnelings_for_https_requests_over_http_proxy {#disable_tunnelings_for_https_requests_over_http_proxy}
```
```suggestion
### disable_tunneling_for_https_requests_over_http_proxy {#disable_tunneling_for_https_requests_over_http_proxy}
```

pufit commented:
Hi, @arthurpassos! Sorry for the delay in the review. Everything looks good to me, except for some minor code style issues: to be consistent with the other clickhouse code, you shouldn't put `);` on a new empty line like here https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-3703ea8d435200e498989803c42205545b9c8ad5f79461529f01c41da800a551R103

arthurpassos committed:
fix setting docs typo

arthurpassos commented:
> Hi, @arthurpassos! Sorry for the delay in the review. Everything looks good to me, except for some minor code style issues: to be consistent with the other clickhouse code, you shouldn't put `);` on a new empty line like here https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-3703ea8d435200e498989803c42205545b9c8ad5f79461529f01c41da800a551R103

Hi @pufit, thanks for looking into this. Does this code style rule also apply to: https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-65bb6a3d5594d1123db3f2de52453405f1425ec60f4e7157d72adba5872798edR38? And what about https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-cbe83137d46d7e2577414410832a81f8df616aaef43924cc36dacb12d05f7b37R26? Can a single parenthesis be in a empty line like https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-c064ca4cd5e4eb5d1014cba51c9781b25457941c7aa0e19a84bdbf4505102823R265?

arthurpassos committed:
move ); up

arthurpassos commented:
Hi @pufit. Can you check if https://github.com/ClickHouse/ClickHouse/pull/55033/commits/e7618ffad47ceb367fe0c7120db45e52fd73e481 addresses your comment?

pufit commented:
> Does this code style rule also apply to: https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-65bb6a3d5594d1123db3f2de52453405f1425ec60f4e7157d72adba5872798edR38? And what about https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-cbe83137d46d7e2577414410832a81f8df616aaef43924cc36dacb12d05f7b37R26?

No, it's only about multiline functions https://clickhouse.com/docs/en/development/style#additional-recommendations

>  Can a single parenthesis be in a empty line like https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-c064ca4cd5e4eb5d1014cba51c9781b25457941c7aa0e19a84bdbf4505102823R265?

`std::tie` is a function, so it should comply with the rule above. As well as function declarations, like here https://github.com/ClickHouse/ClickHouse/pull/55033/files#diff-8c2d29fd63c2e0b955b7cd2f3e607c1936c0114e4021b1bc7699565abfd47feeR22

Sorry for this mess, I will try to edit our style check so it will be easier to deal with in the future. I wish we could use `clang-format`, but unfortunately, this would completely ruin our git history.

arthurpassos committed:
move ) up

arthurpassos commented:
@pufit Just made some more changes, can you check now? Hopefully I caught all of those. If not, can you edit or mark the ones you need me to change?

pufit reviewed (approved):
LGTM!

arthurpassos commented:
Thank you @pufit!
